### PR TITLE
Implement MaskedDistribution

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #520.

Blocked by #574 because we need `expand` method for broadcasting `base_dist` if `mask.ndim > len(batch_shape)`.

 #### TODO

- [x] add more tests for expand

### CODE REVIEW

**Comment by fehiepsi** [comment]
@neerajprad @eb8680 I have also added a couple of bug fixes which are exercised in funsor. This should unblock https://github.com/pyro-ppl/funsor/pull/326. There is only one issue remaining related to [`has_rsample` attribute](https://github.com/pyro-ppl/funsor/blob/master/funsor/distributions.py#L126). Should we add `has_rsample` flag to NumPyro distributions?

File: numpyro/distributions/discrete.py [hunk 1]
```diff
@@ -294,7 +294,7 @@ def variance(self):
 
     @property
     def support(self):
-        return constraints.integer_interval(0, np.shape(self.probs)[-1])
+        return constraints.integer_interval(0, np.shape(self.probs)[-1] - 1)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:297]: this is a bug

File: numpyro/distributions/discrete.py [hunk 2]
```diff
@@ -342,10 +342,10 @@ def variance(self):
 
     @property
     def support(self):
-        return constraints.integer_interval(0, np.shape(self.logits)[-1])
+        return constraints.integer_interval(0, np.shape(self.logits)[-1] - 1)
 
     def enumerate_support(self, expand=True):
-        values = np.arange(self.probs.shape[-1]).reshape((-1,) + (1,) * len(self.batch_shape))
+        values = np.arange(self.logits.shape[-1]).reshape((-1,) + (1,) * len(self.batch_shape))
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:348]: we don't need to use `self.probs` here for CategoricalLogits

**Comment by eb8680** [comment]
> Should we add has_rsample flag to NumPyro distributions?

Might as well, we'll need it if/when we add fancier VI methods to NumPyro via Funsor.
